utils_cv/detection/model.py

```diff
@@ -0,0 +1,283 @@
+            logger = train_one_epoch(
+                self.model,
+                self.optimizer,
+                self.train_dl,
+                self.device,
+                epoch,
+                print_freq=print_freq,
+            )
+            self.losses.append(logger.meters["loss"].median)
+
+            # update the learning rate
+            self.lr_scheduler.step()
+
+            # evaluate
+            self.evaluate(dl=self.train_dl)
+
+    def plot_losses(self, figsize: Tuple[int, int] = (10, 5)) -> None:
+        """ Plot training loss from fitting. """
+        fig = plt.figure(figsize=figsize)
+        ax = fig.add_subplot(111)
+        ax.set_xlim([0, self.epochs - 1])
+        ax.set_xticks(range(0, self.epochs))
+        ax.set_title("Loss over epochs")
+        ax.set_xlabel("epochs")
+        ax.set_ylabel("loss")
+        ax.plot(self.losses)
+
+    def evaluate(self, dl: DataLoader = None) -> CocoEvaluator:
+        """ eval code on validation/test set. """
+        if dl is None:
+            dl = self.test_dl
+        return evaluate(self.model, dl, device=self.device)
+
+    def get_model(self) -> nn.Module:
+        """ returns the model object. """
+        return self.model
+
+    def inference(self, im_path: Union[str, Path]):
+        """ Performs inferencing on an image path. """
+        im = Image.open(im_path)
+        transform = transforms.Compose([transforms.ToTensor()])
+        im = transform(im).cuda()
+        model = self.get_model().eval()  # eval mode
+        with torch.no_grad():
+            pred = model([im])
+        return pred
+
+    def show_detection_vs_ground_truth(self, idx: int, ax: plt.axes) -> None:
+        """ Plots the bounding box predictions and ground truth
+
+        Args:
+            idx: the index of the dataset to visualize
+            ax: axes to draw on
+
+        Returns nothing but plots graph.
+        """
+        im_path = (
+            self.dataset.root
+            / self.dataset.image_folder
+            / self.dataset.ims[idx]
+        )
+        im = cv2.imread(str(im_path))
+
+        # plot prediction boxes
+        pred = self.inference(im_path)
+        pred_labels, pred_boxes = get_bounding_boxes(pred)
+        pred_params = BoundingBoxParams(rect_color=(255, 0, 0), text_size=0)
+        im = plot_boxes(
+            im,
+            pred_boxes,
+            [self.dataset.categories[l] for l in pred_labels],
+            bbox_params=pred_params,
+        )
+
+        # plot ground truth boxes
+        ground_truth_params = BoundingBoxParams(
+            rect_color=(0, 255, 0), text_size=0
+        )
+        boxes, categories, im_path = self.dataset.get_image_features(idx)
+        im = plot_boxes(im, boxes, categories, bbox_params=ground_truth_params)
+
+        # show image
+        ax.set_xticks([])
+        ax.set_yticks([])
+        ax.imshow(im)
+
+    def show_preds(self, ds: Dataset = None, rows: int = 1) -> None:
```

Comment:
[minor] Maybe this is the right thing to do, having eg a helper function called plot_grid. I might myself though repeatedly thinking that this looks too complicated and is not easy to follow. I also somehow don't like the fact that the plotting function executes the model. Seems somehow that this function is doing too much, but could be wrong.

Reply:
yeah, i'll make it so that the plotting function doesn't execute the function. but i find the `plot_grid` util to be nice because multiple functions use it and we can reduce the repeated code.